worlds/mlss/Rules.py

```diff
@@ -0,0 +1,571 @@
+    )
+    add_rule(
+        world.multiworld.get_location(LocationName.PantsShopMomPiranhaFlag1, world.player),
+        lambda state: StateLogic.brooch(state, world.player) or StateLogic.rose(state, world.player),
+    )
+    add_rule(
+        world.multiworld.get_location(LocationName.PantsShopMomPiranhaFlag2, world.player),
+        lambda state: StateLogic.brooch(state, world.player) or StateLogic.rose(state, world.player),
+    )
+    add_rule(
+        world.multiworld.get_location(LocationName.PantsShopMomPiranhaFlag3, world.player),
+        lambda state: StateLogic.brooch(state, world.player) or StateLogic.rose(state, world.player),
+    )
+    add_rule(
+        world.multiworld.get_location(LocationName.BadgeShopMomPiranhaFlag1, world.player),
+        lambda state: StateLogic.brooch(state, world.player) or StateLogic.rose(state, world.player),
+    )
+    add_rule(
+        world.multiworld.get_location(LocationName.BadgeShopMomPiranhaFlag2, world.player),
+        lambda state: StateLogic.brooch(state, world.player) or StateLogic.rose(state, world.player),
+    )
+    add_rule(
+        world.multiworld.get_location(LocationName.BadgeShopMomPiranhaFlag3, world.player),
+        lambda state: StateLogic.brooch(state, world.player) or StateLogic.rose(state, world.player),
+    )
+    add_rule(
+        world.multiworld.get_location(LocationName.ChateauGreenGoblet, world.player),
+        lambda state: StateLogic.brooch(state, world.player) and StateLogic.canDig(state, world.player),
+    )
+    add_rule(
+        world.multiworld.get_location(LocationName.ChateauRedGoblet, world.player),
+        lambda state: StateLogic.brooch(state, world.player) and StateLogic.canMini(state, world.player),
+    )
+
+    add_rule(
+        world.multiworld.get_location(LocationName.GwarharLagoonSpangle, world.player),
+        lambda state: StateLogic.ultra(state, world.player),
+    )
+    add_rule(
+        world.multiworld.get_location(LocationName.GwarharLagoonSpangleRoomBlock, world.player),
+        lambda state: StateLogic.ultra(state, world.player),
+    )
+    if world.options.difficult_logic:
+        add_rule(
+            world.multiworld.get_location(LocationName.GwarharLagoonSpangleReward, world.player),
+            lambda state: StateLogic.canCrash(state, world.player),
+        )
+        add_rule(
+            world.multiworld.get_location(LocationName.BeanstarPieceHermie, world.player),
+            lambda state: StateLogic.canCrash(state, world.player),
+        )
+        if world.options.chuckle_beans != 0:
+            add_rule(
+                world.multiworld.get_location(LocationName.GwarharLagoonPastHermieDigspot, world.player),
+                lambda state: StateLogic.canCrash(state, world.player),
+            )
+
+    if world.options.coins:
+        add_rule(
+            world.multiworld.get_location(LocationName.HoohooMountainBaseBooStatueCaveCoinBlock1, world.player),
```

Comment:
Is there a reason the locations after this point use strings?

Reply:
Purely because the LocationName class was made as part of the original implementation and the coins locations we're added later on in its cycle. I never got around to adding the location names in.